test/publisher_subscriber/PublisherSubscriberInteraction.cpp

```diff
@@ -0,0 +1,144 @@
+    }
+
+    void init_agent(uint16_t port)
+    {
+        switch(transport_)
+        {
+            case UDP_TRANSPORT:
+                agent_.reset(new eprosima::uxr::UDPServer(port));
+                break;
+            case TCP_TRANSPORT:
+                agent_.reset(new eprosima::uxr::TCPServer(port));
+                break;
+        }
+        agent_->run();
+    }
+
+    void check_messages(std::string message, size_t number, uint8_t stream_id_raw)
+    {
+        std::thread publisher_thread(&Client::publish, &publisher_, 1, stream_id_raw, number, message);
+        std::thread subscriber_thread(&Client::subscribe, &subscriber_, 1, stream_id_raw, number, message);
+
+        publisher_thread.join();
+        subscriber_thread.join();
+    }
+
+protected:
+    int transport_;
+    std::unique_ptr<eprosima::uxr::Server> agent_;
+    Client publisher_;
+    Client subscriber_;
+    static const std::string SMALL_MESSAGE;
+};
+
+const std::string PublisherSubscriberInteraction::SMALL_MESSAGE("Hello DDS world!");
+
+INSTANTIATE_TEST_CASE_P(TransportAndLost, PublisherSubscriberInteraction,
+        ::testing::Combine(::testing::Values(UDP_TRANSPORT, TCP_TRANSPORT), ::testing::Values(0.0f, 0.05f, 0.1f)));
+
+TEST_P(PublisherSubscriberInteraction, PubSub1TopicsBestEffort)
+{
+    if(0.0f == std::get<1>(GetParam())) //only without lost
+    {
+        check_messages(SMALL_MESSAGE, 1, 0x01);
+    }
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub10TopicsBestEffort)
+{
+    if(0.0f == std::get<1>(GetParam())) //only without lost
+    {
+        check_messages(SMALL_MESSAGE, 10, 0x01);
+    }
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub1TopicsReliable)
+{
+    check_messages(SMALL_MESSAGE, 1, 0x80);
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub10TopicsReliable)
+{
+    check_messages(SMALL_MESSAGE, 10, 0x80);
+}
+
+/* Fix the non-reliable behavior when messages is higher than the agent history to enable this
+TEST_P(PublisherSubscriberInteraction, PubSub30TopicsReliable)
+{
+    check_messages(SMALL_MESSAGE, 30, 0x80);
+}
+*/
+
+TEST_P(PublisherSubscriberInteraction, PubSub1FragmentedTopic2Parts)
+{
+    std::string message(size_t(publisher_.get_mtu() * 1.5), 'A');
+    check_messages(message, 1, 0x80);
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub3FragmentedTopic2Parts)
+{
+    std::string message(size_t(publisher_.get_mtu() * 1.5), 'A');
+    check_messages(message, 3, 0x80);
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub1FragmentedTopic4Parts)
+{
+    std::string message(size_t(publisher_.get_mtu() * 3.5), 'A');
+    check_messages(message, 1, 0x80);
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub3FragmentedTopic4Parts)
+{
+    std::string message(size_t(publisher_.get_mtu() * 3.5), 'A');
+    check_messages(message, 3, 0x80);
+}
+
+int main(int args, char** argv)
+{
+    ::testing::InitGoogleTest(&args, argv);
+    return RUN_ALL_TESTS();
+}
```

Comment:
We should add a Serial test also. What do you think?

Reply:
Yeah, task 4334 in Redmine